File: Binary二进制/leetcode_762_countPrimeSetBits.py
maxBits = 33
isPrime = [1]*maxBits
isPrime[0] = isPrime[1] = 0
for i in range(2, maxBits):
    if isPrime[i]==0:
        continue
    for j in range(i*2, maxBits, i):
        isPrime[j] = 0

maxNum = 1_000_001
preSum = [0]*(maxNum+1)
for i in range(maxNum):
    preSum[i+1] = preSum[i]+isPrime[i.bit_count()]

class Solution:
    def countPrimeSetBits(self, left: int, right: int) -> int:
        return preSum[right+1]-preSum[left]

if __name__ == "__main__":
    s = Solution()
    left, right = 6, 10
    print(s.countPrimeSetBits(left, right))

File: 02.Binary二进制/leetcode_762_countPrimeSetBits.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int countPrimeSetBits(int left, int right) {
        int maxBits = 33;
        vector<int> isPrime(maxBits, 1);
        isPrime[0] = 0;
        isPrime[1] = 0;
        for (int i=2; i<maxBits; i++){
            if (isPrime[i] == 0) continue;
            for (int j=2*i; j<maxBits; j+=i){
                isPrime[j] = 0;
            }
        }

        int res = 0;
        for (int i=left; i<=right; i++){
            res += isPrime[__builtin_popcount(i)];
        }
        return res;
    }
};

int main(){
    Solution s;
    int left = 6, right = 10;
    cout << s.countPrimeSetBits(left, right) << endl;
}
